cmd/evm/internal/t8ntool/execution.go

```diff
@@ -337,20 +426,101 @@ func (pre *Prestate) Apply(vmConfig vm.Config, chainConfig *params.ChainConfig,
+
+	// MPT pre is the same as the pre state for first conversion block
+	for addr, a := range pre.Pre {
 		statedb.SetCode(addr, a.Code)
 		statedb.SetNonce(addr, a.Nonce)
 		statedb.SetBalance(addr, a.Balance)
 		for k, v := range a.Storage {
 			statedb.SetState(addr, k, v)
 		}
 	}
-	// Commit and re-open to start with a clean state.
-	root, _ := statedb.Commit(0, false)
-	statedb, _ = state.New(root, sdb, nil)
+
+	// If verkle mode started, establish the conversion
+	if verkle {
+		// Commit db an create a snapshot from it.
+		mptRoot, err := statedb.Commit(0, false)
+		if err != nil {
+			panic(err)
+		}
+		rawdb.WritePreimages(mptSdb.DiskDB(), statedb.Preimages())
+		mptSdb.TrieDB().WritePreimages()
+		snaps, err := snapshot.New(snapshot.Config{AsyncBuild: false, CacheSize: 10}, mptSdb.DiskDB(), mptSdb.TrieDB(), mptRoot)
+		if err != nil {
+			panic(err)
+		}
+		if snaps == nil {
+			panic("snapshot is nil")
+		}
+		snaps.Cap(mptRoot, 0)
+
+		// reuse the backend db so that the snapshot can be enumerated
+		sdb := mptSdb // := state.NewDatabaseWithConfig(db, &trie.Config{Verkle: true})
+
+		// Load the conversion status
+		log.Info("loading conversion status", "started", pre.Env.Started)
+		if pre.Env.Started != nil {
+			log.Info("non-nil started", "started", *pre.Env.Started)
+		}
+		sdb.InitTransitionStatus(pre.Env.Started != nil && *pre.Env.Started, pre.Env.Ended != nil && *pre.Env.Ended, mptRoot)
+		log.Info("loading current account address, if available", "available", pre.Env.CurrentAccountAddress != nil)
+		if pre.Env.CurrentAccountAddress != nil {
+			log.Info("loading current account address", "address", *pre.Env.CurrentAccountAddress)
+			sdb.SetCurrentAccountAddress(*pre.Env.CurrentAccountAddress)
+		}
+		if pre.Env.CurrentSlotHash != nil {
+			sdb.SetCurrentSlotHash(*pre.Env.CurrentSlotHash)
+		}
+		if pre.Env.StorageProcessed != nil {
+			sdb.SetStorageProcessed(*pre.Env.StorageProcessed)
+		}
+
+		// start the conversion on the first block
+		log.Info("starting verkle transition?", "in transition", sdb.InTransition(), "transitioned", sdb.Transitioned(), "mpt root", mptRoot)
+		if !sdb.InTransition() && !sdb.Transitioned() {
+			sdb.StartVerkleTransition(mptRoot, mptRoot, chainConfig, chainConfig.PragueTime, mptRoot)
+		}
+
+		statedb, err = state.New(types.EmptyRootHash, sdb, nil)
+		if err != nil {
+			panic(err)
+		}
+
+		// Load verkle tree from prestate
+		var vtr *trie.VerkleTrie
+		switch tr := statedb.GetTrie().(type) {
+		case *trie.VerkleTrie:
+			vtr = tr
+		case *trie.TransitionTrie:
+			vtr = tr.Overlay()
+		default:
+			panic("invalid trie type")
+		}
+
+		// create the vkt, should be empty on first insert
+		for k, v := range pre.VKT {
+			values := make([][]byte, 256)
+			values[k[31]] = make([]byte, 32)
+			copy(values[k[31]], v)
+			vtr.UpdateStem(k.Bytes(), values)
+		}
+
+		root, _ := statedb.Commit(0, false)
+		statedb, err = state.New(root, sdb, snaps)
+		if err != nil {
+			panic(err)
+		}
```

Comment:
snap is indeed stale, and that's a feature: it's the mpt snap that we need for the conversion. I'm adding more comments so that this is less confusing.